src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,188 @@
+package bio.terra.common.retry;
+import bio.terra.common.exception.BadRequestException;
+import bio.terra.common.exception.ConflictException;
+import bio.terra.common.exception.ErrorReportException;
+import bio.terra.common.exception.ForbiddenException;
+import bio.terra.common.exception.InternalServerErrorException;
+
+import bio.terra.common.exception.NotFoundException;
+import bio.terra.common.exception.UnauthorizedException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.api.client.http.HttpStatusCodes;
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.commons.lang3.StringUtils;
+import org.broadinstitute.dsde.workbench.client.sam.ApiException;
+
+import org.broadinstitute.dsde.workbench.client.sam.model.ErrorReport;
+import org.slf4j.Logger;
+
+
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+import org.slf4j.LoggerFactory;
+
+import static java.time.Instant.now;
+
+/**
+ * SamRetry encapsulates logic needed for retrying Sam API calls. The retry function is linear and
+ * all constants are hard-coded, as Sam failures are generally just short periods of down-time.
+ */
+public class SamRetry {
+    private static Logger logger = LoggerFactory.getLogger(SamRetry.class);
+    private static final int MAXIMUM_WAIT_SECONDS = 30;
+    private static final int INITIAL_WAIT_SECONDS = 10;
+    private static final int OPERATION_TIMEOUT_SECONDS = 300;
+    private final Instant operationTimeout;
+    private int retrySeconds;
+
+    private static final ObjectMapper objectMapper = new ObjectMapper();
+
+    SamRetry() {
+        this.operationTimeout = now().plusSeconds(OPERATION_TIMEOUT_SECONDS);
+        this.retrySeconds = INITIAL_WAIT_SECONDS;
+    }
+
+    @VisibleForTesting
+    SamRetry(int timeoutSeconds) {
+        this.operationTimeout = now().plusSeconds(timeoutSeconds);
+        this.retrySeconds = INITIAL_WAIT_SECONDS;
+    }
+
+    @FunctionalInterface
+    public interface SamVoidFunction {
+        void apply() throws ApiException, InterruptedException;
+    }
+
+    @FunctionalInterface
+    public interface SamFunction<R> {
+        R apply() throws ApiException, InterruptedException;
+    }
+
+
+    static <T> T retry(SamFunction<T> function)
```

Comment:
I think this should be public? ditto other static methods?

Reply:
Yes caught that in testing, thank you